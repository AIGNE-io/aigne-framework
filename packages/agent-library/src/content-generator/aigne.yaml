chat_model:
  provider: gemini
  name: gemini-2.5-flash-preview-04-17
  temperature: 0.8
agents:
  - content-generator.yaml
  - structure-planning.yaml
  - evaluation.yaml
  - content-detail-generator.yaml
  - batch-content-detail-generator.yaml
  - page-template-generator.yaml
  - docs-generator.yaml
  - batch-docs-detail-generator.yaml
  - load-sources.mjs
  - save-docs.mjs
