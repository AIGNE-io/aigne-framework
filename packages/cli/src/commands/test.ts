import { spawnSync } from "node:child_process";
import { isAbsolute, resolve } from "node:path";
import { Command } from "commander";

export function createTestCommand(): Command {
  return new Command("test")
    .description("Run tests in the specified agents directory")
    .option(
      "--url, --path <path_or_url>",
      "Path to the agents directory or URL to aigne project",
      ".",
    )
    .action(async (options: { path: string }) => {
      const { path } = options;
      const absolutePath = isAbsolute(path) ? path : resolve(process.cwd(), path);

      spawnSync("node", ["--test"], { cwd: absolutePath, stdio: "inherit" });
    })
    .showHelpAfterError(true)
    .showSuggestionAfterError(true);
}
