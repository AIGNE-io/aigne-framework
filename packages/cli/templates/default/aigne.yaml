chat_model:
  provider: openai
  name: gpt-4o-mini
  temperature: 0.8
agents:
  - chat.yaml
tools:
  - sandbox.js
  - filesystem.yaml
