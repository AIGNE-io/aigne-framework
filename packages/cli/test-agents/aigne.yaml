name: test_aigne_project
description: A test project for the aigne agent
chat_model:
  model: openai:gpt-4o-mini
  temperature: 0.8
agents:
  - chat.yaml
skills:
  - sandbox.js
mcp_server:
  agents:
    - chat.yaml
cli:
  agents:
    - chat.yaml
    - name: component
      description: Manage components of web-smith
      alias: ["comp", "components"]
      commands:
        - url: test-cli-agents/push.yaml
          name: push-components
          alias: ["push", "push-comps"]
        - url: test-cli-agents/pull.yaml
