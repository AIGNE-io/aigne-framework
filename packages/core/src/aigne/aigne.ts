import { z } from "zod";
import { Agent } from "../agents/agent.js";
import { load } from "../loader/index.js";
import { ChatModel } from "../models/chat-model.js";
import { checkArguments, createAccessorArray } from "../utils/type-utils.js";
import { AIGNEContext, type Context } from "./context.js";
import { MessageQueue } from "./message-queue.js";
import type { ContextLimits } from "./usage.js";

/**
 * Options for the AIGNE class.
 */
export interface AIGNEOptions {
  /**
   * The name of the AIGNE instance.
   */
  name?: string;

  /**
   * The description of the AIGNE instance.
   */
  description?: string;

  /**
   * Global model to use for all agents not specifying a model.
   */
  model?: ChatModel;

  /**
   * Skills to use for the AIGNE instance.
   */
  skills?: Agent[];

  /**
   * Agents to use for the AIGNE instance.
   */
  agents?: Agent[];

  /**
   * Limits for the AIGNE instance, such as timeout, max tokens, max invocations, etc.
   */
  limits?: ContextLimits;
}

/**
 * AIGNE is a class that orchestrates multiple agents to build complex AI applications.
 * It serves as the central coordination point for agent interactions, message passing, and execution flow.
 *
 * @example
 * Here's a simple example of how to use AIGNE:
 * {@includeCode ../../test/aigne/aigne.test.ts#example-simple}
 *
 * @example
 * Here's an example of how to use AIGNE with streaming response:
 * {@includeCode ../../test/aigne/aigne.test.ts#example-streaming}
 */
export class AIGNE {
  /**
   * Loads an AIGNE instance from a directory containing an aigne.yaml file and agent definitions.
   * This static method provides a convenient way to initialize an AIGNE system from configuration files.
   *
   * @param path - Path to the directory containing the aigne.yaml file.
   * @param options - Options to override the loaded configuration.
   * @returns A fully initialized AIGNE instance with configured agents and skills.
   */
  static async load(path: string, options?: AIGNEOptions): Promise<AIGNE> {
    const { model, agents, skills, ...aigne } = await load({ path });
    return new AIGNE({
      ...options,
      model: options?.model || model,
      name: options?.name || aigne.name || undefined,
      description: options?.description || aigne.description || undefined,
      agents: agents.concat(options?.agents ?? []),
      skills: skills.concat(options?.skills ?? []),
    });
  }

  /**
   * Creates a new AIGNE instance with the specified options.
   *
   * @param options - Configuration options for the AIGNE instance including name, description, model, and agents.
   */
  constructor(options?: AIGNEOptions) {
    if (options) checkArguments("AIGNE", aigneOptionsSchema, options);

    this.name = options?.name;
    this.description = options?.description;
    this.model = options?.model;
    this.limits = options?.limits;
    if (options?.skills?.length) this.skills.push(...options.skills);
    if (options?.agents?.length) this.addAgent(...options.agents);

    this.initProcessExitHandler();
  }

  /**
   * Optional name identifier for this AIGNE instance.
   */
  name?: string;

  /**
   * Optional description of this AIGNE instance's purpose or functionality.
   */
  description?: string;

  /**
   * Global model to use for all agents that don't specify their own model.
   */
  model?: ChatModel;

  /**
   * Usage limits applied to this AIGNE instance's execution.
   */
  limits?: ContextLimits;

  /**
   * Message queue for handling inter-agent communication.
   */
  readonly messageQueue = new MessageQueue();

  /**
   * Collection of skill agents available to this AIGNE instance.
   * Provides indexed access by skill name.
   */
  readonly skills = createAccessorArray<Agent>([], (arr, name) => arr.find((i) => i.name === name));

  /**
   * Collection of primary agents managed by this AIGNE instance.
   * Provides indexed access by agent name.
   */
  readonly agents = createAccessorArray<Agent>([], (arr, name) => arr.find((i) => i.name === name));

  /**
   * Adds one or more agents to this AIGNE instance.
   * Each agent is attached to this AIGNE instance, allowing it to access the AIGNE's resources.
   *
   * @param agents - One or more Agent instances to add to this AIGNE.
   */
  addAgent(...agents: Agent[]) {
    checkArguments("AIGNE.addAgent", aigneAddAgentArgsSchema, agents);

    for (const agent of agents) {
      this.agents.push(agent);

      agent.attach(this);
    }
  }

  /**
   * Creates a new execution context for this AIGNE instance.
   * Contexts isolate state for different flows or conversations.
   *
   * @returns A new AIGNEContext instance bound to this AIGNE.
   */
  newContext() {
    return new AIGNEContext(this);
  }

  /**
   * Invokes an agent with the specified input and options.
   * This is a shorthand method that creates a new context and immediately invokes an agent.
   *
   * @param args - Arguments passed to the Context's invoke method.
   * @returns The result of the agent invocation.
   *
   * @example
   * Here's a simple example of how to invoke an agent:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-simple}
   *
   * @example
   * Here's an example of how to invoke an agent with streaming response:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-streaming}
   */
  invoke = ((...args: Parameters<Context["invoke"]>) => {
    return new AIGNEContext(this).invoke(...args);
  }) as Context["invoke"];

  /**
   * Publishes a message to the message queue.
   * This is a shorthand method that creates a new context and publishes a message.
   *
   * @param args - Arguments passed to the Context's publish method.
   * @returns The result of the publish operation.
   */
  publish = ((...args) => {
    return new AIGNEContext(this).publish(...args);
  }) as Context["publish"];

  /**
   * Subscribes to messages in the message queue.
   * This allows for receiving messages from agents and other components.
   *
   * @param args - Arguments passed to the MessageQueue's subscribe method.
   * @returns A subscription that can be used to unsubscribe.
   */
  subscribe = ((...args) => {
    return this.messageQueue.subscribe(...args);
  }) as Context["subscribe"];

  /**
   * Unsubscribes from messages in the message queue.
   * This cancels a previous subscription to stop receiving messages.
   *
   * @param args - Arguments passed to the MessageQueue's unsubscribe method.
   */
  unsubscribe = ((...args) => {
    this.messageQueue.unsubscribe(...args);
  }) as Context["unsubscribe"];

  /**
   * Gracefully shuts down the AIGNE instance and all its agents and skills.
   * This ensures proper cleanup of resources before termination.
   *
   * @returns A promise that resolves when shutdown is complete.
   *
   * @example
   * Here's an example of shutdown an AIGNE instance:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-shutdown}
   *
   * @example
   * Here's an example of using async dispose:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-shutdown-using}
   */
  async shutdown() {
    for (const tool of this.skills) {
      await tool.shutdown();
    }
    for (const agent of this.agents) {
      await agent.shutdown();
    }
  }

  /**
   * Asynchronous dispose method for the AIGNE instance.
   *
   * @example
   * Here's an example of using async dispose:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-shutdown-using}
   */
  async [Symbol.asyncDispose]() {
    await this.shutdown();
  }

  /**
   * Initializes handlers for process exit events to ensure clean shutdown.
   * This registers handlers for SIGINT and exit events to properly terminate all agents.
   */
  private initProcessExitHandler() {
    const shutdownAndExit = () => this.shutdown().finally(() => process.exit(0));
    process.on("SIGINT", shutdownAndExit);
    process.on("exit", shutdownAndExit);
  }
}

const aigneOptionsSchema = z.object({
  model: z.instanceof(ChatModel).optional(),
  skills: z.array(z.instanceof(Agent)).optional(),
  agents: z.array(z.instanceof(Agent)).optional(),
});

const aigneAddAgentArgsSchema = z.array(z.instanceof(Agent));
