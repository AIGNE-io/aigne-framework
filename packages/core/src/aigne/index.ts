export * from "./context.js";
export * from "./aigne.js";
export * from "./message-queue.js";
export * from "./usage.js";
