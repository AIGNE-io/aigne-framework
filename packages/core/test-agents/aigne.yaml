name: test_aigne_project
description: A test project for the aigne agent
chat_model:
  name: gpt-4o-mini
  temperature: 0.8
agents:
  - chat.yaml
  - chat-with-prompt.yaml
  - team.yaml
  - image.yaml
  - agents/test-relative-prompt-paths.yaml
skills:
  - sandbox.js
mcp_server:
  agents:
    - chat.yaml
cli:
  agents:
    - chat.yaml
    - test-cli-agents/b.yaml
    - url: test-cli-agents/a.yaml
      name: a-renamed
      description: A agent from a.yaml
      alias: ["a", "a-agent"]
      commands:
        - url: test-cli-agents/a-1.yaml
          commands:
            - url: test-cli-agents/a-1-1.yaml
            - url: test-cli-agents/a-1-2.yaml
              name: a12-renamed
              description: A agent from a-1-2.yaml
        - test-cli-agents/a-2.yaml
