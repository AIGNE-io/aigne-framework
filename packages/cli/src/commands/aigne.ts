import chalk from "chalk";
import yargs from "yargs";
import { AIGNE_CLI_VERSION } from "../constants.js";
import { asciiLogo } from "../utils/ascii-logo.js";
import { createAppCommands } from "./app.js";
import { createConnectCommand } from "./connect.js";
import { createCreateCommand } from "./create.js";
import { createObservabilityCommand } from "./observe.js";
import { createRunCommand } from "./run.js";
import { createServeMCPCommand } from "./serve-mcp.js";
import { createTestCommand } from "./test.js";

export function createAIGNECommand(options?: { aigneFilePath?: string }) {
  console.log(asciiLogo);

  console.log(
    `${chalk.grey("TIPS:")} use ${chalk.greenBright("aigne observe")} to start the observability server.\n`,
  );

  return yargs()
    .scriptName("aigne")
    .usage("CLI for AIGNE framework")
    .version(AIGNE_CLI_VERSION)
    .command(createRunCommand(options))
    .command(createTestCommand(options))
    .command(createCreateCommand())
    .command(createServeMCPCommand(options))
    .command(createObservabilityCommand())
    .command(createConnectCommand())
    .command(createAppCommands())
    .alias("help", "h")
    .alias("version", "v")
    .strict();
}
