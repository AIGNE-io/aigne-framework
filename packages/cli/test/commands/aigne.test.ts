import { expect, test } from "bun:test";
import { createAIGNECommand } from "@aigne/cli/commands/aigne.js";
import { Command } from "commander";

test("aigne command should parse --version correctly", async () => {
  const command = createAIGNECommand();
  expect(command).toBeInstanceOf(Command);

  expect((command as any)["_defaultCommandName"]).toBe("run");
});
