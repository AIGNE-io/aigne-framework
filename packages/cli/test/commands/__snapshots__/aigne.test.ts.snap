// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`aigne command should print help if no any subcommand 1`] = `
[
  [
    
"
     \x1B[38;2;79;172;254m_\x1B[39m    \x1B[38;2;80;170;254m_\x1B[39m\x1B[38;2;81;169;253m_\x1B[39m\x1B[38;2;81;167;253m_\x1B[39m \x1B[38;2;82;166;253m_\x1B[39m\x1B[38;2;83;164;252m_\x1B[39m\x1B[38;2;84;163;252m_\x1B[39m\x1B[38;2;85;161;252m_\x1B[39m \x1B[38;2;86;159;251m_\x1B[39m   \x1B[38;2;86;158;251m_\x1B[39m \x1B[38;2;87;156;251m_\x1B[39m\x1B[38;2;88;155;251m_\x1B[39m\x1B[38;2;89;153;250m_\x1B[39m\x1B[38;2;90;152;250m_\x1B[39m\x1B[38;2;90;150;250m_\x1B[39m
    \x1B[38;2;91;148;249m/\x1B[39m \x1B[38;2;92;147;249m\\\x1B[39m  \x1B[38;2;93;145;249m|\x1B[39m\x1B[38;2;94;144;248m_\x1B[39m \x1B[38;2;95;142;248m_\x1B[39m\x1B[38;2;95;141;248m/\x1B[39m \x1B[38;2;96;139;247m_\x1B[39m\x1B[38;2;97;138;247m_\x1B[39m\x1B[38;2;98;136;247m_\x1B[39m\x1B[38;2;99;134;246m|\x1B[39m \x1B[38;2;99;133;246m\\\x1B[39m \x1B[38;2;100;131;246m|\x1B[39m \x1B[38;2;101;130;245m|\x1B[39m \x1B[38;2;102;128;245m_\x1B[39m\x1B[38;2;103;127;245m_\x1B[39m\x1B[38;2;104;125;244m_\x1B[39m\x1B[38;2;104;123;244m_\x1B[39m\x1B[38;2;105;122;244m|\x1B[39m
   \x1B[38;2;106;120;244m/\x1B[39m \x1B[38;2;107;119;243m_\x1B[39m \x1B[38;2;108;117;243m\\\x1B[39m  \x1B[38;2;108;116;243m|\x1B[39m \x1B[38;2;109;114;242m|\x1B[39m \x1B[38;2;110;112;242m|\x1B[39m  \x1B[38;2;111;111;242m_\x1B[39m\x1B[38;2;112;109;241m|\x1B[39m  \x1B[38;2;113;108;241m\\\x1B[39m\x1B[38;2;113;106;241m|\x1B[39m \x1B[38;2;114;105;240m|\x1B[39m  \x1B[38;2;115;103;240m_\x1B[39m\x1B[38;2;118;103;239m|\x1B[39m
  \x1B[38;2;121;103;237m/\x1B[39m \x1B[38;2;124;103;236m_\x1B[39m\x1B[38;2;127;103;234m_\x1B[39m\x1B[38;2;130;103;233m_\x1B[39m \x1B[38;2;133;103;231m\\\x1B[39m \x1B[38;2;136;103;230m|\x1B[39m \x1B[38;2;139;104;228m|\x1B[39m \x1B[38;2;142;104;227m|\x1B[39m\x1B[38;2;145;104;225m_\x1B[39m\x1B[38;2;148;104;224m|\x1B[39m \x1B[38;2;150;104;222m|\x1B[39m \x1B[38;2;153;104;221m|\x1B[39m\x1B[38;2;156;104;219m\\\x1B[39m  \x1B[38;2;159;104;218m|\x1B[39m \x1B[38;2;162;104;216m|\x1B[39m\x1B[38;2;165;104;215m_\x1B[39m\x1B[38;2;168;104;213m_\x1B[39m\x1B[38;2;171;104;212m_\x1B[39m
 \x1B[38;2;174;104;210m/\x1B[39m\x1B[38;2;177;104;209m_\x1B[39m\x1B[38;2;180;104;207m/\x1B[39m   \x1B[38;2;183;105;206m\\\x1B[39m\x1B[38;2;186;105;204m_\x1B[39m\x1B[38;2;189;105;203m\\\x1B[39m\x1B[38;2;192;105;201m_\x1B[39m\x1B[38;2;195;105;200m_\x1B[39m\x1B[38;2;198;105;198m_\x1B[39m\x1B[38;2;201;105;197m\\\x1B[39m\x1B[38;2;204;105;195m_\x1B[39m\x1B[38;2;207;105;194m_\x1B[39m\x1B[38;2;210;105;192m_\x1B[39m\x1B[38;2;213;105;191m_\x1B[39m\x1B[38;2;215;105;189m|\x1B[39m\x1B[38;2;218;105;188m_\x1B[39m\x1B[38;2;221;105;186m|\x1B[39m \x1B[38;2;224;105;185m\\\x1B[39m\x1B[38;2;227;106;183m_\x1B[39m\x1B[38;2;230;106;182m|\x1B[39m\x1B[38;2;233;106;180m_\x1B[39m\x1B[38;2;236;106;179m_\x1B[39m\x1B[38;2;239;106;177m_\x1B[39m\x1B[38;2;242;106;176m_\x1B[39m\x1B[38;2;245;106;174m_\x1B[39m\x1B[38;2;248;106;173m|\x1B[39m

\x1B[36mvxx.xx.xx\x1B[39m


aigne <command> [options]

Commands:
  aigne run [path] [entry-agent]  Run AIGNE for the specified path
  aigne eval [path] [agent]       Evaluate AIGNE for the specified path
  aigne test                      Run tests in the specified agents directory
  aigne create [path]             Create a new aigne project with agent config files
  aigne serve-mcp                 Serve the agents in the specified directory as a MCP server (streamable http)
  aigne observe                   Start the observability server
  aigne doc-smith                 Generate and maintain project docs — powered by agents.  [aliases: docsmith, doc]
  aigne web-smith                 Generate and maintain project website pages — powered by agents.  [aliases: websmith, web]
  aigne hub <command>             Manage AIGNE Hub connections
  aigne deploy                    Deploy an aigne application

Options:
  -h, --help     Show help  [boolean]
  -v, --version  Show version number  [boolean]
"
,
  ],
  [],
  [
    "Not enough non-option arguments: got 0, need at least 1",
  ],
]
`;
